### ISSUE TITLE
Issues in `replay buffer`

### ISSUE THREAD

**Issue by hyeok9855** [GFNOrg/torchgfn#201]
Here I note several issues in the current `ReplayBuffer`, including the points raised by @younik in #193.

1. **Types of objects**, (#193): Currently, `ReplayBuffer` can take `Trajectories`, `Transitions` (which are inherited from `Container`), and `tuple[States, States]` (which is not). This makes the overall code a bit more messy. Possible solutions are as follows: 
  a. Make `State` inherit from `Container`, and let the `ReplayBuffer` take `Container` objects.
  b. Make subclasses of `ReplayBuffer` depending on the objects
2. **Supporting terminal-state-only Buffer**: I think the most popular form of the `ReplayBuffer` is just storing a terminal state x (with reward) as in [Max's code](https://github.com/maxwshen/gflownet), and sample a training trajectory with backward policy. This is preferred since 1) it significantly reduces memory usage (vs. Trajectory-based), and 2) the backward sampling gives more diverse learning signals, which I believe helps training. We need to support this, but it needs to be considered along with the issue 1 above.
3. **Device** (minor): Currently, the device of objects in the buffer follows that of env. Maybe we should support options for both CPU and cuda, since they have different pros and cons, e.g, storing as cuda tensor is faster if we use cuda, but it consumes more GPU memory. ***BTW, there exist some minor errors in `ReplayBuffer` when cuda is enabled. We need to check.***

I hope to discuss enough before tackling those!

**Comment by josephdviviano** [comment]
I'm also happy to help with this :)

**Comment by josephdviviano** [comment]
Thanks -- re this - I'm wondering whether we can simply use the `ReplayBuffer` implementation in the torchrl distribution and inherit all of their functionality. This would obviously be it's own PR and might not be possible if too many decisions we made elsewhere in the repo are incompatible (but at the very least this is a good reference implementation for us).

https://github.com/pytorch/rl/tree/main/torchrl/data/replay_buffers

**Comment by younik** [comment]
Reusing TorchRL ReplayBuffer is a very good idea. They decouple sampling from the replay buffer, which is a great design choice.

I will look at the challenges to migrating to it more in depth

**Comment by hyeok9855** [comment]
Personally, I don't want to include torchrl, only to use the buffer. I will first try refactoring without it. Let's see.

**Comment by hyeok9855** [comment]
Regarding prioritization, the current implementation of `PrioritizedReplayBuffer` uses the reward (=priority score) just to sort the samples in the buffer and then drop low-reward ones when the buffer is full.

However, I think the core of prioritization is prioritized sampling, as in [here](https://pytorch.org/rl/stable/tutorials/rb_tutorial.html#prioritized-replay-buffers).

Here are some ideas to improve regarding the prioritization: 
1. Separate the storage and sampler, similar to the torchrl. Storage manages how to do when it is full (e.g., FIFO, sort using priority and drop tails ...), and the sampler manages how to sample, including prioritized sampling.
2. Allow any scalar metrics for prioritization (e.g., reward, [training loss](https://arxiv.org/abs/2410.01432), etc.)

**Comment by hyeok9855** [comment]
Can anyone explain why do we need `terminating_states` ([here](https://github.com/GFNOrg/torchgfn/blob/a59b5810f619bcfb6dbb83058c672937c9fc44f5/src/gfn/containers/replay_buffer.py#L42)) in the replay buffer?

**Comment by saleml** [comment]
Thanks @hyeok9855 for the comments
- I agree with you. `torchrl` might be an overkill.
- `terminating_states` is for the flow matching loss. Unfortunately, this loss requires two input objects, unlike all other losses. I also learned that this is a design choice of this library, and flow matching does *not* have to be implemented this way. Do you have a strong opinion on this @hyeok9855 ?

**Comment by hyeok9855** [comment]
Thanks @saleml.

I've just checked the flow matching loss. I don't have a strong opinion, but I wonder if we could improve in this way:

Can we simply add a flag indicating a state is terminal or not (maybe the existence of `_log_reward`?) and then calculate everything inside of `flow_matching_loss` (rather than using additional `reward_matching_loss`)?

**Comment by saleml** [comment]
I'm definitely ok with your suggestion. Some authors refrain from using the "reward_matching_loss" term altogether.
@josephdviviano , I'd love your opinion here too.

**Comment by josephdviviano** [comment]
@hyeok9855 if I am reading this discussion correctly, based on recent PRs, the major only piece that needs to be added is the terminal state-only buffer. Does that check out for you as well?

**Comment by hyeok9855** [comment]
> the major only piece that needs to be added is the terminal state-only buffer. Does that check out for you as well?

Yes!
